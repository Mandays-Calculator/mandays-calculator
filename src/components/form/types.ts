import type { ReactElement, ReactNode } from "react";

export type BaseInputProps<Type = object> = Type & {
  name?: string;
  label?: string | ReactNode;
  error?: boolean;
  helperText?: string | ReactNode;
  isFastField?: boolean;
  readOnly?: boolean;
};

export type WithInputControllerType =
  | "text"
  | "date-picker"
  | "dropzone"
  | "pin-code"
  | "autocomplete"
  | "checkbox-group"
  | "date-time-picker"
  | "number-input";

export type BaseInputPropsWithAdornment<Type = object> = Type & {
  name: string;
  label?: string;
  endAdornment?: ReactElement;
  readOnly?: boolean;
  placeholder?: string;
};

export interface WithInputControllerValueProp {
  value?: string;
  selected?: string;
  startDate?: string;
  endDate?: string;
}

type FormValues = {
  [key: string]: any;
};

export type FormErrors = {
  [key in keyof FormValues]?: string;
};
