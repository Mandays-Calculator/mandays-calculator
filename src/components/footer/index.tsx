export { default as Footer } from "./Footer";
export { StyledFooterContainer, StyledFooterText } from "./styles";
