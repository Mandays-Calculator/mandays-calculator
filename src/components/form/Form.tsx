import type { PropsWithChildren, ReactElement } from "react";
import type { FormikInstance } from "formik";
import { FormikProvider } from "formik";
interface FormProps<FormValues extends object> extends PropsWithChildren {
  instance: FormikInstance<FormValues>;
}

export const Form = <FormValues extends object>(props: FormProps<FormValues>): ReactElement => {
  const { instance, children, ...rest } = props;
  return (
    <FormikProvider value={instance}>
      <form
        {...rest}
        onSubmit={instance.handleSubmit}
      >
        {children}
      </form>
    </FormikProvider>
  );
};

export default Form;
