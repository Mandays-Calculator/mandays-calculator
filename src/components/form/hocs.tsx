import type { ComponentType, ReactElement } from "react";
import type { FormikValues } from "formik";
import type {
  BaseInputProps,
  WithInputControllerType,
  WithInputControllerValueProp,
} from "./types";
import type { ReactDatePickerFunctionParams } from "../global-types/app";

import { useMemo } from "react";
import { useFormikContext, getIn, Field, FastField } from "formik";
import { isArray } from "lodash";

type WrappedInputProps<Type> = Type &
  BaseInputProps & {
    name: string;
  };
type WithInputController<Type extends object> = (props: WrappedInputProps<Type>) => ReactElement;

export const withInputController = <Type extends FormikValues>(
  WrappedInput: ComponentType<Type>,
  inputType: WithInputControllerType = "text"
): WithInputController<Type> => {
  return (props: WrappedInputProps<Type>): ReactElement => {
    const { name, isFastField, ...rest } = props;
    const form = useFormikContext<Type>();

    const fieldValue = form.values[name] || getIn(form.values, name);
    const fieldError = form.errors[name] || getIn(form.errors, name);
    const fieldTouched = form.touched[name] || getIn(form.touched, name);
    const hasError = Boolean(fieldError) && Boolean(fieldTouched);

    const handleInputNumberChange = (_: any, value: number): void => {
      form.setFieldValue(name, value);
    };

    const handleDateChange = (date: ReactDatePickerFunctionParams): void => {
      form.setFieldValue(name, date);
    };

    const handleDropzoneChange = (files: File[]): void => {
      form.setFieldValue(name, files);
    };

    const handlePinCodeChange = (pinCode: string): void => {
      form.setFieldValue(name, pinCode);
    };

    const handleCheckBoxChanges = (checkBoxName: string, checked: boolean): void => {
      if (checked) {
        form.setFieldValue(name, [checkBoxName, ...(form.values[name] ?? [])]);
      } else {
        const checkBoxData: string[] = [...form.values[name]];
        const index = checkBoxData.indexOf(checkBoxName);

        if (index !== -1) {
          checkBoxData.splice(index, 1);
          form.setFieldValue(name, checkBoxData);
        }
      }
    };

    const handleAutoComplete = (data: Type): void => {
      form.setFieldValue(name, data);
    };

    const getInputChange = ():
      | typeof handleDateChange
      | typeof handleDropzoneChange
      | typeof handlePinCodeChange
      | typeof handleCheckBoxChanges
      | typeof handleAutoComplete
      | typeof handleInputNumberChange
      | typeof form.handleChange => {
      switch (inputType) {
        case "date-picker":
          return handleDateChange;
        case "dropzone":
          return handleDropzoneChange;
        case "pin-code":
          return handlePinCodeChange;
        case "checkbox-group":
          return handleCheckBoxChanges;
        case "autocomplete":
          return handleAutoComplete;
        case "number-input":
          return handleInputNumberChange;
        default:
          return form.handleChange(name);
      }
    };

    const getInputValueProp = (): WithInputControllerValueProp => {
      if (inputType === "date-picker") {
        if (isArray(fieldValue)) {
          return {
            startDate: fieldValue[0],
            endDate: fieldValue[1],
          };
        }
        return {
          selected: fieldValue,
        };
      }

      return {
        value: fieldValue,
      };
    };

    const MemoriedField = useMemo(() => (isFastField ? FastField : Field), [isFastField]);
    return (
      <MemoriedField
        component={WrappedInput}
        name={name}
        error={hasError}
        {...getInputValueProp()}
        onChange={getInputChange()}
        {...(rest as Type)}
      />
    );
  };
};
