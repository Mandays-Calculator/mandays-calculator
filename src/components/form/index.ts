export * from "./checkbox";
export * from "./checkbox-group";
export * from "./text-field";
export * from "./date-picker";
