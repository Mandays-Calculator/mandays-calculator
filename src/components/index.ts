export * from "./form/text-field";
export * from "./form/Form";
export * from "./title";
