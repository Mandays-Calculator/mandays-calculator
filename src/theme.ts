import { createTheme } from "@mui/material/styles";

const theme = createTheme({
  palette: {
    primary: {
      main: "#2C8ED1",
      dark: "#414145",
      light: "#ADDEFF",
      contrastText: "#F9F4F4",
    },
    secondary: {
      main: "#D54147",
      light: "#EAF3F4",
    },
    background: {
      default: "#FEFEFE",
      paper: "#F5F5F5",
    },
    success: {
      main: "#85DD74",
    },
    warning: {
      main: "#ffd633",
    },
    error: {
      main: "#FF4545",
    },
  },
  breakpoints: {
    values: {
      xs: 0,
      sm: 600,
      md: 1024,
      lg: 1280,
      xl: 1367,
    },
  },
  typography: {
    fontFamily: "Montserrat, sans-serif",
    htmlFontSize: 14,
    h1: {
      fontSize: "2rem",
      fontWeight: 600,
      color: "#414145",
    },
    h2: {
      fontSize: "1.5rem",
      fontWeight: 500,
      color: "#414145",
    },
    body1: {
      fontSize: "1rem",
      color: "#414145",
    },
  },
  spacing: (factor: number) => `${0.5 * factor}rem`,
  components: {
    MuiLink: {
      styleOverrides: {
        root: {
          color: "#2c8ed1",
          textDecoration: "none",
          "&:hover": {
            color: "#4699D2",
          },
        },
      },
    },
    MuiCssBaseline: {
      styleOverrides: {
        html: {
          fontSize: "16px",
          "@media (min-width:1280px)": {
            fontSize: "14px",
          },
        },
      },
    },
  },
});

export default theme;
