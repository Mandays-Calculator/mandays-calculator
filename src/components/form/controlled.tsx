import { CheckBox, CheckboxGroup, DatePicker, SearchSelect, TextField, NumberInput } from ".";
import DateTimePicker from "./date-time-picker/DateTimePicker";
import { withInputController } from "./hocs";
import Select from "./select/Select";
import TextArea from "./text-area/TextArea";

export const ControlledTextField = withInputController(TextField);
export const ControlledCheckBox = withInputController(CheckBox);
export const ControlledTextArea = withInputController(TextArea);
export const ControlledCheckBoxGroup = withInputController(
  CheckboxGroup,
  "checkbox-group"
) as typeof CheckboxGroup;
export const ControlledDatePicker = withInputController(
  DatePicker,
  "date-picker"
);
export const ControlledSelect = withInputController(Select);
export const ControlledSearchSelect = withInputController(
  SearchSelect,
  "autocomplete"
) as typeof SearchSelect;
export const ControlledDateTimePicker = withInputController(
  DateTimePicker,
  "date-time-picker"
);
export const ControlledNumberInput = withInputController(NumberInput, "number-input");
