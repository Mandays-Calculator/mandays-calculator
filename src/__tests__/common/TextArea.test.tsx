import { render } from '@testing-library/react';
import { ThemeProvider } from "@mui/material";
import theme from "~/theme";
import { TextArea } from "~/components/form";


describe('TextArea Component', () => {
  test('renders without errors', () => {
    const {getByTestId, asFragment} = render(
      <TextArea 
        name="test" 
        label="Text Area Label"
      />
    );
    const textarea = getByTestId("text-area-component");
    expect(textarea).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test('handles input and triggers events', () => {
    const { getByTestId, asFragment} = render(<TextArea name="test" />);
    const textarea = getByTestId("text-area-component");
    expect(textarea).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test('render with error', () => {
    const { getByTestId, asFragment, unmount} = render(
      <ThemeProvider theme={theme}>
        <TextArea 
          name="test" 
          error
        />
      </ThemeProvider>
    );
    const textarea = getByTestId("text-area-component");
    expect(textarea).toBeInTheDocument();
    expect(textarea).toHaveStyle({
      borderColor: theme.palette.error.main,
    });
    expect(asFragment()).toMatchSnapshot();
    unmount();
  })
});