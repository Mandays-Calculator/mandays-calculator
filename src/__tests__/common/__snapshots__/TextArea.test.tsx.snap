// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea Component handles input and triggers events 1`] = `
<DocumentFragment>
  <div
    class="MuiStack-root css-hwnj0i-MuiStack-root"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
    >
      <textarea
        aria-label="test"
        class="css-bmb4ke"
        data-testid="text-area-component"
        id="test"
        placeholder="Please Input"
        rows="1"
        style="height: 0px; overflow: hidden;"
      />
      <textarea
        aria-hidden="true"
        class="css-bmb4ke"
        readonly=""
        style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding-top: 0px; padding-bottom: 0px; width: 100%;"
        tabindex="-1"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`TextArea Component render with error 1`] = `
<DocumentFragment>
  <div
    class="MuiStack-root css-5u2gm5-MuiStack-root"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
    >
      <textarea
        aria-label="test"
        class="css-1egmxn1"
        data-testid="text-area-component"
        id="test"
        placeholder="Please Input"
        rows="1"
        style="height: 0px; overflow: hidden;"
      />
      <textarea
        aria-hidden="true"
        class="css-1egmxn1"
        readonly=""
        style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding-top: 0px; padding-bottom: 0px; width: 100%;"
        tabindex="-1"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`TextArea Component renders without errors 1`] = `
<DocumentFragment>
  <div
    class="MuiStack-root css-hwnj0i-MuiStack-root"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      >
        Text Area Label
      </p>
      <textarea
        aria-label="test"
        class="css-bmb4ke"
        data-testid="text-area-component"
        id="test"
        placeholder="Please Input"
        rows="1"
        style="height: 0px; overflow: hidden;"
      />
      <textarea
        aria-hidden="true"
        class="css-bmb4ke"
        readonly=""
        style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding-top: 0px; padding-bottom: 0px; width: 100%;"
        tabindex="-1"
      />
    </div>
  </div>
</DocumentFragment>
`;
